buildSrc/src/main/kotlin/io/spine/internal/gradle/publish/Publications.kt

```diff
@@ -0,0 +1,203 @@
+        maven {
+            url = project.uri(target)
+            credentials {
+                username = credentials?.username
+                password = credentials?.password
+            }
+        }
+    }
+}
+
+/**
+ * A publication for a typical Java project.
+ *
+ * In Gradle, in order to publish something somewhere one should create a publication.
+ * A publication has a name and consists of one or more artifacts plus information about
+ * those artifacts – the metadata.
+ *
+ * An instance of this class represents [MavenPublication] named "mavenJava". It is generally
+ * accepted that a publication with this name contains a Java project published to one or
+ * more Maven repositories.
+ *
+ * By default, only a jar with the compilation output of `main` source set and its
+ * metadata files are published. Other artifacts are specified through the
+ * [constructor parameter][jars]. Please, take a look on [specifyArtifacts] for additional info.
+ *
+ * @param artifactId
+ *         a name that a project is known by.
+ * @param jars
+ *         list of artifacts to be published along with the compilation output.
+ * @param destinations
+ *         Maven repositories to which the produced artifacts will be sent.
+ * @see <a href="https://docs.gradle.org/current/userguide/publishing_maven.html#publishing_maven:publications">
+ *       Maven Publish Plugin | Publications</a>
+ */
+internal class MavenJavaPublication(
+    artifactId: String,
+    private val jars: Set<TaskProvider<Jar>>,
+    destinations: Set<Repository>,
+) : PublicationHandler(artifactId, destinations) {
+
+    /**
+     * Creates a new "mavenJava" [MavenPublication] in the given project.
+     */
+    override fun handlePublications(project: Project) {
+        project.publications.create<MavenPublication>("mavenJava") {
+            assignMavenCoordinates(project)
+            specifyArtifacts(project, jars)
+        }
+    }
+}
+
+/**
+ * Specifies which artifacts this [MavenPublication] will contain.
+ *
+ * A typical Maven publication contains:
+ *
+ *  1. Jar archives. For example: compilation output, sources, javadoc, etc.
+ *  2. Maven metadata file that has ".pom" extension.
+ *  3. Gradle metadata file that has ".module" extension.
+ *
+ *  Metadata files contain information about a publication itself, its artifacts and their
+ *  dependencies. Presence of ".pom" file is mandatory for publication to be consumed by
+ *  `mvn` build tool itself or other build tools that understand Maven notation (Gradle, Ivy).
+ *  Presence of ".module" is optional, but useful when a publication is consumed by Gradle.
+ *
+ * @see <a href="https://maven.apache.org/pom.html">Maven – POM Reference</a>
+ * @see <a href="https://docs.gradle.org/current/userguide/publishing_gradle_module_metadata.html">
+ *      Understanding Gradle Module Metadata</a>
+ */
+private fun MavenPublication.specifyArtifacts(project: Project, jars: Set<TaskProvider<Jar>>) {
```

Comment:
Can this method go to `MavenJavaPublication` ? It seems to be used only there.

Reply:
It used to be there. I specifically moved it out because of 1) too deep nesting 2) aiming to keep the `MavenJavaPublication` smaller so that a reader has the better focus on what the class does. 

Makes sense?